File: java/filament/src/main/java/io/github/erkko68/filament/Stream.java
package io.github.erkko68.filament;

import io.github.erkko68.filament.internal.NativeRegistry;
import java.lang.ref.Cleaner;

public class Stream {
    private long mNativeObject;

    private Stream(long nativeStream) {
        mNativeObject = nativeStream;
    }

    public static class Builder {
        private final long mNativeBuilder;
        private final Cleaner.Cleanable mCleanable;

        public Builder() {
            mNativeBuilder = nCreateBuilder();
            mCleanable = NativeRegistry.registerForCleanup(this, new BuilderCleanup(mNativeBuilder));
        }

        public Builder stream(long nativeTexture) {
            nBuilderStream(mNativeBuilder, nativeTexture);
            return this;
        }

        public Builder width(int width) {
            nBuilderWidth(mNativeBuilder, width);
            return this;
        }

        public Builder height(int height) {
            nBuilderHeight(mNativeBuilder, height);
            return this;
        }

        public Stream build(Engine engine) {
            long nativeObject = nBuilderBuild(mNativeBuilder, engine.getNativeObject());
            if (nativeObject == 0) throw new IllegalStateException("Couldn't create Stream");
            return new Stream(nativeObject);
        }

        private static class BuilderCleanup implements Runnable {
            private final long mNativeObject;
            BuilderCleanup(long nativeObject) { mNativeObject = nativeObject; }
            @Override public void run() { nDestroyBuilder(mNativeObject); }
        }
    }

    public StreamType getStreamType() {
        return StreamType.values()[nGetStreamType(getNativeObject())];
    }

    public enum StreamType {
        NATIVE,
        ACQUIRED
    }

    public long getNativeObject() {
        if (mNativeObject == 0) {
            throw new IllegalStateException("Calling method on destroyed Stream");
        }
        return mNativeObject;
    }

    void clearNativeObject() {
        mNativeObject = 0;
    }

    private static native long nCreateBuilder();
    private static native void nDestroyBuilder(long nativeBuilder);
    private static native void nBuilderStream(long nativeBuilder, long nativeTexture);
    private static native void nBuilderWidth(long nativeBuilder, int width);
    private static native void nBuilderHeight(long nativeBuilder, int height);
    private static native long nBuilderBuild(long nativeBuilder, long nativeEngine);
    private static native int nGetStreamType(long nativeStream);
}

File: kotlin/filament/src/jvmMain/kotlin/io/github/erkko68/filament/Stream.kt
package io.github.erkko68.filament

import io.github.erkko68.filament.jni.Stream as JniStream

actual class Stream(val nativeStream: JniStream) {
    actual class Builder actual constructor() {
        private val jni = JniStream.Builder()

        actual fun width(width: Int): Builder {
            jni.width(width)
            return this
        }

        actual fun height(height: Int): Builder {
            jni.height(height)
            return this
        }

        actual fun stream(streamSource: Any): Builder {
            jni.stream(streamSource)
            return this
        }

        actual fun build(engine: Engine): Stream =
            Stream(jni.build(engine.nativeEngine))
    }

    actual enum class StreamType {
        NATIVE, ACQUIRED
    }

    actual fun getStreamType(): StreamType = StreamType.values()[nativeStream.streamType.ordinal]

    actual fun setDimensions(width: Int, height: Int) {
        nativeStream.setDimensions(width, height)
    }

    actual fun getTimestamp(): Long = nativeStream.timestamp
}
